modules/common/src/main/java/StringUtils.java

```diff
@@ -1,62 +1,77 @@
 package lila.common.base;
 
 public class StringUtils {
+    private static final char[] DIGITS = {
+        '0' , '1' , '2' , '3' , '4' , '5' ,
+        '6' , '7' , '8' , '9' , 'a' , 'b' ,
+        'c' , 'd' , 'e' , 'f'
+    };
+
     public static String safeJsonString(String s) {
         char[] sArr = s.toCharArray();
+        StringBuilder sb = null;
         int len = sArr.length;
-        int sizeGuess = len > 0 && sArr[0] <= '~' ? 1 : 6;
-        StringBuilder sb = new StringBuilder(sizeGuess * len + 2);
-
-        sb.append('"');
-        for (char c : sArr) {
-            boolean safe = c >= ' ' && c <= '~';
-            if (safe) switch(c) {
+        int safeIdx = 0;
```

Comment:
Since the index itself has not been checked (only the previous one), maybe find a better name.

Reply:
hmm, the safe range is `[safeIdx, i)`, so the range is empty when `safeIdx == i`, as is the case at the start.

Reply:
ah, makes sense viewing it this way